File: src/main/java/io/gittul/domain/user/dto/UserDetailResponse.java
package io.gittul.domain.user.dto;

import io.gittul.domain.follow.entity.UserFollow;
import io.gittul.domain.user.entity.User;
import io.gittul.domain.user.entity.UserDetails;

import java.util.List;

public record UserDetailResponse(
        Long userId,
        String nickname,
        String email,
        String profileImageUrl,
        Integer followerCount,
        Integer followingCount,
        List<UserProfileResponse> followings,
        Integer likedCount,
        List<String> interests
) {

    public static UserDetailResponse of(User user) {
        UserDetails details = user.getDetails();

        int likeCount = details.getPosts().stream()
                .mapToInt(post -> post.getLikes().size())
                .sum();

        return new UserDetailResponse(
                user.getUserId(),
                user.getUserName(),
                user.getEmail(),
                user.getProfileImage(),
                details.getFollowers().size(),
                details.getFollowings().size(),
                details.getFollowings().stream()
                        .map(UserFollow::getFollowee)
                        .map(UserProfileResponse::of)
                        .toList(),
                likeCount,
                details.getInterests().stream()
                        .map(userInterest -> userInterest.getTag().getTagName())
                        .toList()
        );
    }
}

File: app/src/main/kotlin/io/gittul/app/domain/user/dto/UserDetailResponse.kt
package io.gittul.app.domain.user.dto

import io.gittul.core.domain.user.entity.User

data class UserDetailResponse(
    val userId: Long,
    val nickname: String,
    val email: String,
    val profileImageUrl: String?,
    val followerCount: Int,
    val followingCount: Int,
    val followings: List<UserProfileResponse>,
    val likedCount: Int,
    val interests: List<String>
) {
    companion object {
        fun of(user: User): UserDetailResponse {
            val details = user.details

            val likeCount = details.threads.stream()
                .mapToInt { it.likes.size }
                .sum()

            return UserDetailResponse(
                user.userId,
                user.userName,
                user.email,
                user.profileImage,
                details.followers.size,
                details.followings.size,
                details.followings.stream()
                    .map<User> { it.followee }
                    .map<UserProfileResponse> { UserProfileResponse.of(it) }
                    .toList(),
                likeCount,
                details.interests.stream()
                    .map<String> { it.tag.tagName }
                    .toList()
            )
        }
    }
}
